### ISSUE TITLE
GLES2 error: initialize: Directional shadow framebuffer status invalid on macOS

### ISSUE THREAD

**Issue by robgraeber** [godotengine/godot#26570]
**Godot version:** 3.1 beta 10

**OS/device including version:** OS X 10.12 AMD Radeon Pro 455

**Issue description:** Hey I noticed when restarting the editor sometimes the old editor doesn't fully close and instead hangs forever. I noticed it more yesterday when I was trying out the GLES2 VS GLES3 settings, but can't reproduce it 100%. Are there logs files somewhere?

**Steps to reproduce:** Hard to reproduce, happens sporadically when doing "Quit to project list" or switching the video driver.

**Minimal reproduction project:** In this case I was testing with Dodge the Creeps.

<img width="629" alt="screen shot 2019-03-04 at 12 56 58 am" src="https://user-images.githubusercontent.com/2387719/53722210-da5ad080-3e19-11e9-85e9-e094abab12bb.png">

**Comment by fian46** [comment]
i can confirm that happened on ubuntu 18.04 too sometimes

**Comment by akien-mga** [comment]
Can you run Godot from a temrinal and see if any errors/backtraces are displayed when this happens?

**Comment by bruvzg** [comment]
#24206 and #24669 duplicate?

**Comment by akien-mga** [comment]
Probably, if so should be fixed in the master branch / beta 11 / RC 1.

**Comment by akien-mga** [comment]
Assuming duplicate of #24669 and thus fixed. Please comment if you still reproduce it in the next build, and I'll reopen it.

**Comment by robgraeber** [comment]
@akien-mga it's still happening on rc2. Here's the log when running Godot from cli (switching from GLES3 to GLES2 on dodge the creeps:

```
Running: /Applications/Godot.app/Contents/MacOS/Godot --path /Users/bobby/Downloads/godot-demo-projects-master/2d/dodge_the_creeps --remote-debug 127.0.0.1:6007 --allow_focus_steal_pid 46001 --position 480,90
arguments
0: /Applications/Godot.app/Contents/MacOS/Godot
1: --path
2: /Users/bobby/Downloads/godot-demo-projects-master/2d/dodge_the_creeps
3: --remote-debug
4: 127.0.0.1:6007
5: --allow_focus_steal_pid
6: 46001
7: --position
8: 480,90
Current path: /Applications/Godot.app/Contents/Resources
OpenGL ES 3.0 Renderer: AMD Radeon Pro 455 OpenGL Engine
WARNING: cleanup: ObjectDB Instances still exist!
   At: core/object.cpp:2092.
arguments
0: /Applications/Godot.app/Contents/MacOS/Godot
1: --path
2: /Users/bobby/Downloads/godot-demo-projects-master/2d/dodge_the_creeps
3: -e
4: res://Main.tscn
Current path: /Applications/Godot.app/Contents/Resources
OpenGL ES 2.0 Renderer: AMD Radeon Pro 455 OpenGL Engine
ERROR: initialize: Directional shadow framebuffer status invalid
   At: drivers/gles2/rasterizer_scene_gles2.cpp:3355.  
```

**Comment by akien-mga** [comment]
Do you get the same error when running a GLES2 project directly?

**Comment by akien-mga** [comment]
And does it run if you remove `default_env.tres`?

**Comment by akien-mga** [comment]
And finally, could you run Godot with the `--verbose` argument to have more detailed output (should show the OpenGL errors).

Should be something like `/Applications/Godot.app/Contents/MacOS/Godot --verbose -e --path /Users/bobby/Downloads/godot-demo-projects-master/2d/dodge_the_creeps`

**Comment by bruvzg** [comment]
I do get `ERROR: initialize: Directional shadow framebuffer status invalid` error when running Godot with the GLES2 driver (with both existing GLES3 project and newly created GLES2 project), but can't reproduce crash or freeze.

@robgraeber Check `Console.app`->`User Reports` for Godot crash logs.

**Comment by akien-mga** [comment]
> *Edit:* same GLES2 error is present in beta 1 and onwards.

Does that mean that 3.1 alpha 5 worked, or you just tested down to beta 1?

**Comment by bruvzg** [comment]
> Does that mean that 3.1 alpha 5 worked, or you just tested down to beta 1?

I just tested all betas. And I do not find anything not working with any, it's just printing this error on the start. It's probably completely irrelevant to this issue or #26912.

**Comment by bruvzg** [comment]
Done some more test and shadows are NOT working in GLES2 (I assume it should work if it's trying to setup directional shadow framebuffer):

GLES 3 (shadow ok):
![Screenshot 2019-03-11 at 22 28 22](https://user-images.githubusercontent.com/7645683/54155599-24ad0480-444d-11e9-9076-1132db2bcf58.png)

GLES 2 (beta 1, 7 and 9, no shadow, everything is dark):
![Screenshot 2019-03-11 at 22 29 07](https://user-images.githubusercontent.com/7645683/54155614-2f679980-444d-11e9-8de3-7960229f72de.png)

GLES 2 (beta 10+, no shadow, everything is covered in stripes):
![Screenshot 2019-03-11 at 22 28 06](https://user-images.githubusercontent.com/7645683/54155852-a6049700-444d-11e9-949d-41882dbd2b57.png)

GLES2 (after #26928, fully broken):

![Screenshot 2019-03-11 at 22 37 20](https://user-images.githubusercontent.com/7645683/54156127-422e9e00-444e-11e9-86c2-111a1bc9cc54.png)

<details>
  <summary>Console output</summary>

```
OpenGL ES 2.0 Renderer: AMD Radeon HD - FirePro D300 OpenGL Engine
1: #version 120
2: #define USE_GLES_OVER_GL
3: #define USE_TEXTURE_RECT
4:
5: #ifdef USE_GLES_OVER_GL
6: #define lowp
7: #define mediump
8: #define highp
9: #else
10: precision highp float;
11: precision highp int;
12: #endif
13:
14: uniform highp mat4 projection_matrix;
15: /* clang-format on */
16:
17: #ifndef USE_GLES_OVER_GL
18: #ifdef GL_EXT_shader_texture_lod
19: #extension GL_EXT_shader_texture_lod : enable
20: #define texture2DLod(img, coord, lod) texture2DLodEXT(img, coord, lod)
21: #define textureCubeLod(img, coord, lod) textureCubeLodEXT(img, coord, lod)
22: #endif
23: #endif // !USE_GLES_OVER_GL
24:
25: #ifdef GL_ARB_shader_texture_lod
26: #extension GL_ARB_shader_texture_lod : enable
27: #endif
28:
29: #if !defined(GL_EXT_shader_texture_lod) && !defined(GL_ARB_shader_texture_lod)
30: #define texture2DLod(img, coord, lod) texture2D(img, coord, lod)
31: #define textureCubeLod(img, coord, lod) textureCube(img, coord, lod)
32: #endif
33:
34: vec2 select2(vec2 a, vec2 b, bvec2 c) {
35: 	vec2 ret;
36:
37: 	ret.x = c.x ? b.x : a.x;
38: 	ret.y = c.y ? b.y : a.y;
39:
40: 	return ret;
41: }
42:
43: vec3 select3(vec3 a, vec3 b, bvec3 c) {
44: 	vec3 ret;
45:
46: 	ret.x = c.x ? b.x : a.x;
47: 	ret.y = c.y ? b.y : a.y;
48: 	ret.z = c.z ? b.z : a.z;
49:
50: 	return ret;
51: }
52:
53: vec4 select4(vec4 a, vec4 b, bvec4 c) {
54: 	vec4 ret;
55:
56: 	ret.x = c.x ? b.x : a.x;
57: 	ret.y = c.y ? b.y : a.y;
58: 	ret.z = c.z ? b.z : a.z;
59: 	ret.w = c.w ? b.w : a.w;
60:
61: 	return ret;
62: }
63:
64: highp vec4 texel2DFetch(highp sampler2D tex, ivec2 size, ivec2 coord) {
65: 	float x_coord = float(2 * coord.x + 1) / float(size.x * 2);
66: 	float y_coord = float(2 * coord.y + 1) / float(size.y * 2);
67:
68: 	return texture2DLod(tex, vec2(x_coord, y_coord), 0.0);
69: }
70:
71: #ifndef USE_GLES_OVER_GL
72: highp mat4 transpose(highp mat4 src) {
73: 	return mat4(
74: 			vec4(src[0].x, src[1].x, src[2].x, src[3].x),
75: 			vec4(src[0].y, src[1].y, src[2].y, src[3].y),
76: 			vec4(src[0].z, src[1].z, src[2].z, src[3].z),
77: 			vec4(src[0].w, src[1].w, src[2].w, src[3].w));
78: }
79: #endif
80:
81: uniform highp mat4 modelview_matrix;
82: uniform highp mat4 extra_matrix;
83: attribute highp vec2 vertex; // attrib:0
84: attribute vec4 color_attrib; // attrib:3
85: attribute vec2 uv_attrib; // attrib:4
86:
87: #ifdef USE_SKELETON
88: attribute highp vec4 bone_indices; // attrib:6
89: attribute highp vec4 bone_weights; // attrib:7
90: #endif
91:
92: #ifdef USE_INSTANCING
93:
94: attribute highp vec4 instance_xform0; //attrib:8
95: attribute highp vec4 instance_xform1; //attrib:9
96: attribute highp vec4 instance_xform2; //attrib:10
97: attribute highp vec4 instance_color; //attrib:11
98:
99: #ifdef USE_INSTANCE_CUSTOM
100: attribute highp vec4 instance_custom_data; //attrib:12
101: #endif
102:
103: #endif
104:
105: #ifdef USE_SKELETON
106: uniform highp sampler2D skeleton_texture; // texunit:-3
107: uniform highp ivec2 skeleton_texture_size;
108: uniform highp mat4 skeleton_transform;
109: uniform highp mat4 skeleton_transform_inverse;
110: #endif
111:
112: varying vec2 uv_interp;
113: varying vec4 color_interp;
114:
115: uniform highp vec2 color_texpixel_size;
116:
117: #ifdef USE_TEXTURE_RECT
118:
119: uniform vec4 dst_rect;
120: uniform vec4 src_rect;
121:
122: #endif
123:
124: uniform highp float time;
125:
126: #ifdef USE_LIGHTING
127:
128: // light matrices
129: uniform highp mat4 light_matrix;
130: uniform highp mat4 light_matrix_inverse;
131: uniform highp mat4 light_local_matrix;
132: uniform highp mat4 shadow_matrix;
133: uniform highp vec4 light_color;
134: uniform highp vec4 light_shadow_color;
135: uniform highp vec2 light_pos;
136: uniform highp float shadowpixel_size;
137: uniform highp float shadow_gradient;
138: uniform highp float light_height;
139: uniform highp float light_outside_alpha;
140: uniform highp float shadow_distance_mult;
141:
142: varying vec4 light_uv_interp;
143: varying vec2 transformed_light_uv;
144: varying vec4 local_rot;
145:
146: #ifdef USE_SHADOWS
147: varying highp vec2 pos;
148: #endif
149:
150: const bool at_light_pass = true;
151: #else
152: const bool at_light_pass = false;
153: #endif
154:
155: /* clang-format off */
156:
157:
158: /* clang-format on */
159:
160: vec2 select(vec2 a, vec2 b, bvec2 c) {
161: 	vec2 ret;
162:
163: 	ret.x = c.x ? b.x : a.x;
164: 	ret.y = c.y ? b.y : a.y;
165:
166: 	return ret;
167: }
168:
169: void main() {
170:
171: 	vec4 color = color_attrib;
172: 	vec2 uv;
173:
174: #ifdef USE_INSTANCING
175: 	mat4 extra_matrix_instance = extra_matrix * transpose(mat4(instance_xform0, instance_xform1, instance_xform2, vec4(0.0, 0.0, 0.0, 1.0)));
176: 	color *= instance_color;
177: 	vec4 instance_custom = instance_custom_data;
178:
179: #else
180: 	mat4 extra_matrix_instance = extra_matrix;
181: 	vec4 instance_custom = vec4(0.0);
182: #endif
183:
184: #ifdef USE_TEXTURE_RECT
185:
186: 	if (dst_rect.z < 0.0) { // Transpose is encoded as negative dst_rect.z
187: 		uv = src_rect.xy + abs(src_rect.zw) * vertex.yx;
188: 	} else {
189: 		uv = src_rect.xy + abs(src_rect.zw) * vertex;
190: 	}
191:
192: 	vec4 outvec = vec4(0.0, 0.0, 0.0, 1.0);
193:
194: 	// This is what is done in the GLES 3 bindings and should
195: 	// take care of flipped rects.
196: 	//
197: 	// But it doesn't.
198: 	// I don't know why, will need to investigate further.
199:
200: 	outvec.xy = dst_rect.xy + abs(dst_rect.zw) * select(vertex, vec2(1.0, 1.0) - vertex, lessThan(src_rect.zw, vec2(0.0, 0.0)));
201:
202: 	// outvec.xy = dst_rect.xy + abs(dst_rect.zw) * vertex;
203: #else
204: 	vec4 outvec = vec4(vertex.xy, 0.0, 1.0);
205:
206: 	uv = uv_attrib;
207: #endif
208:
209: 	{
210: 		vec2 src_vtx = outvec.xy;
211: 		/* clang-format off */
212:
213:
214: 		/* clang-format on */
215: 	}
216:
217: #if !defined(SKIP_TRANSFORM_USED)
218: 	outvec = extra_matrix_instance * outvec;
219: 	outvec = modelview_matrix * outvec;
220: #endif
221:
222: 	color_interp = color;
223:
224: #ifdef USE_PIXEL_SNAP
225: 	outvec.xy = floor(outvec + 0.5).xy;
226: 	// precision issue on some hardware creates artifacts within texture
227: 	// offset uv by a small amount to avoid
228: 	uv += 1e-5;
229: #endif
230:
231: #ifdef USE_SKELETON
232:
233: 	// look up transform from the "pose texture"
234: 	if (bone_weights != vec4(0.0)) {
235:
236: 		highp mat4 bone_transform = mat4(0.0);
237:
238: 		for (int i = 0; i < 4; i++) {
239: 			ivec2 tex_ofs = ivec2(int(bone_indices[i]) * 2, 0);
240:
241: 			highp mat4 b = mat4(
242: 					texel2DFetch(skeleton_texture, skeleton_texture_size, tex_ofs + ivec2(0, 0)),
243: 					texel2DFetch(skeleton_texture, skeleton_texture_size, tex_ofs + ivec2(1, 0)),
244: 					vec4(0.0, 0.0, 1.0, 0.0),
245: 					vec4(0.0, 0.0, 0.0, 1.0));
246:
247: 			bone_transform += b * bone_weights[i];
248: 		}
249:
250: 		mat4 bone_matrix = skeleton_transform * transpose(bone_transform) * skeleton_transform_inverse;
251:
252: 		outvec = bone_matrix * outvec;
253: 	}
254:
255: #endif
256:
257: 	uv_interp = uv;
258: 	gl_Position = projection_matrix * outvec;
259:
260: #ifdef USE_LIGHTING
261:
262: 	light_uv_interp.xy = (light_matrix * outvec).xy;
263: 	light_uv_interp.zw = (light_local_matrix * outvec).xy;
264:
265: 	transformed_light_uv = (mat3(light_matrix_inverse) * vec3(light_uv_interp.zw, 0.0)).xy; //for normal mapping
266:
267: #ifdef USE_SHADOWS
268: 	pos = outvec.xy;
269: #endif
270:
271: 	local_rot.xy = normalize((modelview_matrix * (extra_matrix_instance * vec4(1.0, 0.0, 0.0, 0.0))).xy);
272: 	local_rot.zw = normalize((modelview_matrix * (extra_matrix_instance * vec4(0.0, 1.0, 0.0, 0.0))).xy);
273: #ifdef USE_TEXTURE_RECT
274: 	local_rot.xy *= sign(src_rect.z);
275: 	local_rot.zw *= sign(src_rect.w);
276: #endif
277:
278: #endif
279: }
280:
281: /* clang-format off */
282:
ERROR: _display_error_with_code: CanvasShaderGLES2: Vertex shader compilation failed:
ERROR: 0:26: '' : syntax error: #extension must always be before any non-preprocessor tokens

   At: drivers/gles2/shader_gles2.cpp:129.
ERROR: get_current_version: Method/Function Failed, returning: __null
   At: drivers/gles2/shader_gles2.cpp:288.
ERROR: bind: Condition ' !version ' is true. returned: false
   At: drivers/gles2/shader_gles2.cpp:88.
ERROR: initialize: Directional shadow framebuffer status invalid
   At: drivers/gles2/rasterizer_scene_gles2.cpp:3355.
```

</details>

**Comment by akien-mga** [comment]
> GLES2 (after #26928, fully broken):

Fixed in #26932.

**Comment by JohnCWakley** [comment]
I am still seeing this error in 3.1-stable using GLES2:

E 0:00:00:0050   Directional shadow framebuffer status invalid
  <C Source>     drivers/gles2/rasterizer_scene_gles2.cpp:3355 @ initialize()

Godot 3.1-stable
macOS 10.14.3

**Comment by robgraeber** [comment]
For the freezing issue on mac, I don't see any logs in the Console/User Reports and verbose mode didn't have any new information. It happens on rc3 too. Not sure if it's related, but when I close the editor in general I get this error:
```
ERROR: cleanup: There are still MemoryPool allocs in use at exit!
   At: core/pool_vector.cpp:70.
```

**Comment by JohnCWakley** [comment]
I haven't experienced any freezing, and didn't notice that error. I will try again soon and see if I am getting that as well...

**Comment by JohnCWakley** [comment]
Running my project from Terminal:

ERROR: initialize: Directional shadow framebuffer status invalid
   At: drivers/gles2/rasterizer_scene_gles2.cpp:3355.

I didn't see any other errors, or any freezing.

**Comment by JohnCWakley** [comment]
I built from the 3.1-stable tag with target=debug
macOS: 10.14.3 (2017 Macbook Pro)

After running my project:

OpenGL ES 2.0 Renderer: AMD Radeon Pro 560 OpenGL Engine
ERROR: initialize: Directional shadow framebuffer status invalid: 
   At: drivers/gles2/rasterizer_scene_gles2.cpp:3355.
WARNING: _update_root_rect: Font oversampling does not work in 'Viewport' stretch mode, only '2D'.
   At: scene/main/scene_tree.cpp:1241.
GameManager: _ready
args: {host:127.0.0.1, port:3000}
Start: _ready
_on_Button_Play_Solo_pressed
GameManager: startSoloPlay
WARNING: cleanup: ObjectDB Instances still exist!
   At: core/object.cpp:2092.
ERROR: cleanup: There are still MemoryPool allocs in use at exit!
   At: core/pool_vector.cpp:70.

**Comment by SuperDIMMaX** [comment]
some problem - Windows 10, Radeon 4500m, Godot 3.any version, OpenGL ES 2/3 - no 3d shadows

ERROR: initialize: Condition ' status != 0x8CD5 ' is true. Continuing..:
   At: drivers/gles3/rasterizer_scene_gles3.cpp:4996
ERROR: initialize: Directional shadow framebuffer status invalid
   At: drivers/gles3/rasterizer_scene_gles3.cpp:5021

or

ERROR: initialize: Directional shadow framebuffer status invalid
   At: drivers/gles2/rasterizer_scene_gles2.cpp:3355

**Comment by ImperatorS79** [comment]
Message originally posted in #19896, but I think it is better here:

Hello, I also have a HD graphics 3000, on windows 7, OpenGL 3.1.
I tried to launch the Wesnoth's Godot demo, it worked in Godot 3.1 beta 11, but not in stable.
Here are the console outputs:
- beta 11 :
```
OpenGL ES 2.0 Renderer: Intel(R) HD Graphics 3000
ERROR: initialize: Directional shadow framebuffer status invalid
   At: drivers/gles2/rasterizer_scene_gles2.cpp:3250
```
- stable when godot start:
```
OpenGL ES 2.0 Renderer: Intel(R) HD Graphics 3000
ERROR: initialize: Directional shadow framebuffer status invalid
   At: drivers/gles2/rasterizer_scene_gles2.cpp:3355
```
- stable after launching project
```
ERROR: is_viable: Error initializing GLAD
   At: drivers/gles3/rasterizer_gles3.cpp:142
```
It does not work in rc3 either, and in rc1 I only have the GLAD error at the beginning and I cannot even try to launch the project.

**Comment by dirkk0** [comment]
I can reproduce this on a MBP Mac OSX High Sierra (Intel Iris Graphics 6100), when I switch between GLES2 and 3.
On GLES2 the error is:
E 0:00:00:0132   Directional shadow framebuffer status invalid
  <C Source>     drivers/gles2/rasterizer_scene_gles2.cpp:3196 @ initialize()
Happens in both runtime and editor.

In GLES3 the shadows show.

It might be worth noting that
a) the error doesn't show always, most of the time the shadows just don't show and
b) when I set Max Distance to something between 2 and 5, I do see shadows:
https://www.dropbox.com/s/wyzsp27fa5kw8oq/Screenshot%202019-03-25%2011.54.16.png?dl=0

**Comment by hiulit** [comment]
- Godot 3.1-stable
- MacBookPro mid 2015
- 2D project

```
OpenGL ES 2.0 Renderer: Intel Iris Pro OpenGL Engine
```

```
E 0:00:00:0062   Directional shadow framebuffer status invalid
  <C Source>     drivers/gles2/rasterizer_scene_gles2.cpp:3355 @ initialize()
```

**Comment by ArdaE** [comment]
Same problem here on macOS 10.13.6 (MacBookPro Retina, mid-2012, NVIDIA GeForce GT 650M, Intel HD Graphics 4000).

Happens even in an empty scene with just Node2D or Control at the top level, in a freshly created project. I've tried playing around with some related-looking project settings to no avail.

**Comment by clayheaton** [comment]
@akien-mga 

Here's an example project: https://github.com/clayheaton/JJCopyExportBug

(If this is a different bug or should be referenced elsewhere, please let me know.)

Is it possible this affects the GLES2 export to HTML5? I have a project that exports flawlessly to HTML5 from GLES3 and runs fine in the editor or as an exported app when in GLES2, but doesn't work properly when exported to HTML5 when GLES2 is active. 

(Actually, I get the WebGL Context lost issue when I try to resize with the GLES3 export in Chrome. Maybe related to https://github.com/godotengine/godot/issues/16050? It won't allow me to load the game in the browser if I have the JavaScript console open but works fine when that's not open.)

Here's a screenshot of the errors in Chrome when I do an HTML5 export from GLES2:

![Screen Shot 2019-07-10 at 10 29 49 AM](https://user-images.githubusercontent.com/205364/60977956-5b613180-a2fe-11e9-9ffb-84f3ac68119b.png)
 
This is what it is supposed to look like and this screen is from an HTML5 export from GLES3, running in Chrome.

![Screen Shot 2019-07-10 at 10 35 51 AM](https://user-images.githubusercontent.com/205364/60978075-9499a180-a2fe-11e9-93f5-076af2f38675.png)

**Comment by akien-mga** [comment]
Yeah, there are definitely some issues in the GLES2 setup on platforms with more restrictive implementations (macOS, iOS, WebGL and some older Intel IGPs on Windows). There were lots of changes prior to the 3.1 release around this code to fix similar issues, but it looks like there is still work to do to make the GLES2 setup fully compliant with the lowest end.

**Comment by clayheaton** [comment]
For what it's worth, I tried exporting from GLES2 on Windows (on a Macbook Pro) and encountered the same failure described above.
